compositor_pipeline/src/pipeline/rtp/rtp_input/rtcp_sync.rs

```diff
@@ -0,0 +1,209 @@
+use std::{
+    sync::{Arc, RwLock},
+    time::{Duration, Instant},
+};
+
+use tracing::{debug, warn};
+
+use crate::pipeline::rtp::rtp_input::rollover_state::RolloverState;
+
+const POW_2_32: f64 = (1i64 << 32) as f64;
+
+#[derive(Debug)]
+/// State that should be shared between different RTP tracks to use for synchronization.
+pub(crate) struct RtpNtpSyncPoint {
+    sync_point: Instant,
+    /// First 32 bytes represent seconds, last 32 bytes fraction of the second.
+    /// Represents NTP time of sync point
+    ntp_time: RwLock<Option<u64>>,
+}
+
+impl RtpNtpSyncPoint {
+    pub fn new(sync_point: Instant) -> Arc<Self> {
+        Self {
+            sync_point,
+            ntp_time: RwLock::new(None),
+        }
+        .into()
+    }
+
+    fn ntp_time_to_pts(&self, ntp_time: u64) -> Duration {
+        let sync_point_ntp_time = self.ntp_time.read().unwrap().unwrap_or(0) as i128;
+        let ntp_time_diff_secs = (ntp_time as i128 - sync_point_ntp_time) as f64 / POW_2_32;
+        Duration::try_from_secs_f64(ntp_time_diff_secs).unwrap_or_else(|err| {
+            warn!(%err, "NTP time from before sync point");
+            Duration::ZERO
+        })
+    }
+
+    /// sr_ntp_time - NTP time from SenderReport
+    /// rtp_timestamp - rtp timestamp from SenderReport (represents sr_ntp_time)
+    /// reference_rtp_timestamp - rtp timestamp of some reference RTP packet
+    /// reference_pts - pts(duration from sync_point without buffer) representing above packet
+    fn ensure_sync_info(
+        &self,
+        sr_ntp_time: u64,
+        sr_rtp_timestamp: u32,
+        cmp_rtp_timestamp: u32,
+        cmp_pts: Duration,
+        clock_rate: u32,
+    ) {
+        {
+            let guard = self.ntp_time.read().unwrap();
+            if guard.is_some() {
+                return;
+            }
+        }
+
+        let mut guard = self.ntp_time.write().unwrap();
```

Comment:
This might break if some PTS are before rollover and some after; however, this will only be called once at the beginning of the stream, so it's extremely unlikely. I might look into that in follow-up, but don't want to focus on this right now.